_posts/2018-03-28-building-coreboot-on-minnowboard.md

```diff
@@ -0,0 +1,244 @@
+dd 0984d6012h ; Checksum
+dd 000000001h ; Loader Version
+dd 000000001h ; Platform ID
+dd 00000cbd0h ; Data size
+dd 00000cc00h ; Total size
+dd 000000000h ; reserved
+dd 000000000h ; reserved
+dd 000000000h ; reserved
+dd 000000000h
+dd 0000000a1h
+dd 000020001h
+dd 000000321h
+```
+
+details of it's structure are confidential. However for coreboot we need to
+provide it as comma separated list of values in C-style format, like this:
+
+```C
+0x000000001,
+0x00000021d,
+0x008122013,
+0x000030672,
+0x0a565e78c,
+0x000000001,
+0x000000002,
+0x00000cbd0,
+0x00000cc00,
+0x000000000,
+0x000000000,
+0x000000000,
+0x000000000,
+0x0000000a1,
+```
+
+We can easly download and convert it using simple script:
+
+```bash
+wget 'http://intel.ly/2yStb2e' -O MM-firmware.zip
+unzip MM-firmware.zip
+cd MinnowBoard_MAX-0.97-Binary.Objects/Vlv2MiscBinariesPkg/Microcode
+for f in *.inc ; do
+    name=${f%.inc}
+    awk '{gsub( /h.*$/, "", $2 ); print "0x" $2 ","; }' $f \
+        > $name.h
+done
+cp *.h ../../../
+```
+
+FSP (Firmware Support Package)
+---
+
+FSP is binary package from Intel dedicated to firmware developers containing
+most important platform initialization (including IP). However there is non-FSP
+version of coreboot for Bay Trail however it won't work without MCU code which
+is confidential (available for trusted vendors) so we won't cover this option.
+
+Intel sites redirect to https://github.com/IntelFsp/FSP (branch BayTrail) where
+you can find it. We are interested mostly in BayTrailFspBinPkg/FspBin/BayTrailFSP.fd
+which is file we have to provide to coreboot. FSP can be configured using BCT
+(Binary Configuration Tools), which is optional — needed to enable Secure boot
+and FastBoot. The tool is available only as Windows application, but works well
+using Wine too.
+
+In the same package you can find VgaBIOS which can be used too if you want to
+use graphic card: BayTrailFspBinPkg/Vbios/Vga.dat
+
+ME region
+---------
+
+The simplest way to prevent ME region corruption is to read ROM layout from
+original firmware image. In `utils/ifdtool` of coreboot source tree we can
+find program for reading layout from ROM image. The image can be taken from
+firmware package or read using flashrom, using SPI interface:
+
+```sh
+sudo flashrom -pdediprog -r minnow.rom
+```
+
+Note that you may need to adjust `-p` option according to used SPI programmer.
+Use `-pinternal` if you use MinnowBoard's internal programmer.
+
+```sh
+~/code/coreboot> cd util/ifdtool/
+~/code/coreboot/util/ifdtool> make
+~/code/coreboot/util/ifdtool> ./ifdtool -f ../../mb.layout minnow.rom
+File minnow.bin is 8388608 bytes
+Wrote layout to ../../mb.layout
+~/code/coreboot/util/ifdtool> cat ../../mb.layout
+00000000:00000fff fd
+00400000:007fffff bios
+00001000:003fffff me
+00000000:00000fff gbe
+```
+
+While flashing coreboot we should inform flashrom only to write `bios` region
+
+```sh
+sudo flashrom -l mb.layout -ibios -pdediprog -w build/coreboot.rom
+```
```

Comment:
@BartekLew it would be good to add couple information:

1. The layout may change with firmware that means layout file should not be reused for different firmware version.
2. To recover from broken system one should flash original firmware, read layout and then flash coreboot.

Is this correct to provide options without space? `-ibios`?

Reply:
1, 2: :+1: 

Yes, it's usually valid and I do it like this in flashrom. However I added spaces that to avoid confusion.